src/ir/auth_logic/ast.h

```diff
@@ -93,32 +93,25 @@ class BaseFact {
 class CanSay;
 class Fact;
 
-// CanSay corresponds to an expression of the form <principal> canSay Fact
 class CanSay {
  public:
-  explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
-      : principal_(principal), fact_(std::move(fact)) {}
+  explicit CanSay(Principal principal) : principal_(std::move(principal)) {}
   const Principal& principal() const { return principal_; }
-  const Fact* fact() const { return fact_.get(); }
 
  private:
   Principal principal_;
-  std::unique_ptr<Fact> fact_;
 };
 
-// Fact corresponds to either a base fact or a an expression of the form
-// <principal> canSay <Fact>
 class Fact {
  public:
-  // FactVariantType gives the different forms of Facts. Client code
-  // should use this type to traverse these forms. This type may be changed in
-  // the future.
-  using FactVariantType = std::variant<BaseFact, std::unique_ptr<CanSay>>;
-  explicit Fact(FactVariantType value) : value_(std::move(value)) {}
-  const FactVariantType& GetValue() const { return value_; }
+  Fact(std::forward_list<CanSay> cansay, BaseFact basefact)
+      : cansay_(std::move(cansay)), basefact_(std::move(basefact)) {}
+  const std::forward_list<CanSay>& cansay() const { return cansay_; }
+  const BaseFact& basefact() const { return basefact_; }
 
  private:
-  FactVariantType value_;
+  std::forward_list<CanSay> cansay_;
```

Comment:
In general, in C++, `std::vector` is preferred over `std::forward_list`. The reason for this is that usually they are about equally easy to work with, but `std::vector` allocates its memory in one big, contiguous chunk, which is friendlier to caches and memory management.

In this case, however, you've got a pretty good simplicity argument for using `std::forward_list`, as you wish to push elements onto the front of your list. Just wanted to point out that for the future, though.

Reply:
Noted!